shimmy/meltingpot_compatibility.py

```diff
@@ -0,0 +1,193 @@
+
+        Args:
+            agent (AgentID): agent
+
+        Returns:
+            observation_space: spaces.Space
+        """
+        observation_space = utils.remove_world_observations_from_space(
+            utils.spec_to_space(self._env.observation_spec()[0])  # type: ignore
+        )
+        return observation_space
+
+    @functools.lru_cache(maxsize=None)
+    def action_space(self, agent: AgentID) -> gymnasium.spaces.Space:
+        """action_space.
+
+        Get the action space from the underlying meltingpot substrate.
+
+        Args:
+            agent (AgentID): agent
+
+        Returns:
+            action_space: spaces.Space
+        """
+        action_space = utils.spec_to_space(self._env.action_spec()[0])
+        return action_space
+
+    def state(self) -> np.ndarray:
+        return self._env.observation()
+
+    def reset(
+            self,
+            seed: Optional[int] = None,
+            return_info: bool = False,
+            options: Optional[dict] = None,
+    ) -> ObsDict:
+        """reset.
+
+        Resets the environment.
+
+        Args:
+            seed: the seed to reset the environment with
+            options: the options to reset the environment with
+
+        Returns:
+            (observation, info)
+        """
+        timestep = self._env.reset()
+        self.agents = self.possible_agents[:]
+        self.num_cycles = 0
+        return utils.timestep_to_observations(timestep)
+
+    def step(
+            self, actions: ActionDict
+    ) -> Tuple[
+        ObsDict, Dict[str, float], Dict[str, bool], Dict[str, bool], Dict[str, dict]
+    ]:
+        """step.
+
+        Steps through the environment.
+
+        Args:
+            action: action to step through the environment with
+
+        Returns:
+            (observation, reward, terminated, truncated, info)
+        """
+        actions = [actions[agent] for agent in self.agents]
+        timestep = self._env.step(actions)
+        rewards = {
+            agent: timestep.reward[index] for index, agent in enumerate(self.agents)
+        }
+        self.num_cycles += 1
+        termination = timestep.last()
+        terminations = {agent: termination for agent in self.agents}
+        truncation = self.num_cycles >= self.max_cycles
+        truncations = {agent: truncation for agent in self.agents}
+        infos = {agent: {} for agent in self.agents}
+        if termination or truncation:
+            self.agents = []
+
+        observations = utils.timestep_to_observations(timestep)
+        return observations, rewards, terminations, truncations, infos
+
+    def close(self):
+        self._env.close()
+
+    def render(self) -> None | np.ndarray:
+        """Renders the environment.
+
+        Returns:
+            The rendering of the environment, depending on the render mode
+        """
+        rgb_arr = self.state()[0]['WORLD.RGB']
+        if self.render_mode == 'human':
+            plt.cla()
```

Comment:
True, I think we should avoid import matplotlib if possible. I'm not sure how it's done in meltingpot, but if it's not possible to not use matplotlib, I think that's still fine. We can simply make it a dependency for meltingpot only.

@elliottower is it possible to use PIL instead?

Reply:
I’m not familiar with PIL but from some googling looks doable. https://stackoverflow.com/questions/57316491/how-to-convert-matplotlib-figure-to-pil-image-object-without-saving-image
Something like this? Or do you mean to avoid matplotlib entirely? Is it because plt is slow? And I’ve on seen pettingzoo games use pygame not PIL, is there a consensus on one or the other?

Reply:
We generally try to avoid imports of large libraries, especially if we're not using their main functionality. Since we only want to display images, MPL is a bit overkill. That said, PIL is also a bit of a large library, and so is PyGame. I think the consensus for one over the other has thus far just been 'most of the other libraries tend not to use MPL, so let's not use that'. There may be other reasons that I don't know of.

Reply:
Makes sense, will look into it today.